src/Sarif/Comparers/ArtifactSortingComparer.cs

```diff
@@ -0,0 +1,139 @@
+            {
+                return 1;
+            }
+
+            int compareResult = 0;
+            compareResult = MessageSortingComparer.Instance.Compare(left.Description, right.Description);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = ArtifactLocationSortingComparer.Instance.Compare(left.Location, right.Location);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = left.ParentIndex.CompareTo(right.ParentIndex);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = left.Offset.CompareTo(right.Offset);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = left.Length.CompareTo(right.Length);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = left.Roles.CompareTo(right.Roles);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = string.Compare(left.MimeType, right.MimeType);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = ArtifactContentSortingComparer.Instance.Compare(left.Contents, right.Contents);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = string.Compare(left.Encoding, right.Encoding);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            compareResult = string.Compare(left.SourceLanguage, right.SourceLanguage);
+            if (compareResult != 0)
+            {
+                return compareResult;
+            }
+
+            if (!object.ReferenceEquals(left.Hashes, right.Hashes))
+            {
+                if (left.Hashes == null)
+                {
+                    return -1;
+                }
+
+                if (right.Hashes == null)
+                {
+                    return 1;
+                }
+
+                compareResult = left.Hashes.Count.CompareTo(right.Hashes.Count);
+                if (compareResult != 0)
+                {
+                    return compareResult;
+                }
+
+                for (int i = 0; i < left.Hashes.Count; i++)
+                {
+                    compareResult = string.Compare(left.Hashes.ElementAt(i).Key, right.Hashes.ElementAt(i).Key);
+                    if (compareResult != 0)
+                    {
+                        return compareResult;
+                    }
+
+                    compareResult = string.Compare(left.Hashes.ElementAt(i).Value, right.Hashes.ElementAt(i).Value);
```

Comment:
>left.Hashes.ElementAt(i) [](http://example.com/codeflow?start=51&length=24)

is this an n^2 access pattern? ouch.

Reply:
fixed. new implementation in ComparerHelper.CompareDictionary(...)